src/launch/mod.rs

```diff
@@ -261,3 +263,54 @@ pub struct TdxCapabilities {
 
     pub cpuid_configs: Vec<CpuidConfig>,
 }
+
+/// Manually create the wrapper for KVM_MEMORY_ENCRYPT_OP since `kvm_ioctls` doesn't
+/// support `.encrypt_op` for vcpu fds
+use vmm_sys_util::*;
+ioctl_iowr_nr!(
+    KVM_MEMORY_ENCRYPT_OP,
+    kvm_bindings::KVMIO,
+    0xba,
+    std::os::raw::c_ulong
+);
+
+pub struct TdxVcpu<'a> {
+    pub fd: &'a mut kvm_ioctls::VcpuFd,
+}
+
+impl<'a> TdxVcpu<'a> {
+    pub fn init(&self, hob_address: u64) -> Result<(), TdxError> {
+        let mut cmd = Cmd {
+            id: linux::CmdId::InitVcpu as u32,
+            flags: 0,
+            data: hob_address as *const u64 as _,
+            error: 0,
+            _unused: 0,
+        };
+        let ret = unsafe { ioctl::ioctl_with_mut_ptr(self.fd, KVM_MEMORY_ENCRYPT_OP(), &mut cmd) };
+        if ret < 0 {
+            // can't return `ret` because it will just return -1 and not give the error
+            // code. `cmd.error` will also just be 0.
+            return Err(TdxError::from(errno::Error::last()));
+        }
+        Ok(())
+    }
+}
+
+impl<'a> TryFrom<&'a mut kvm_ioctls::VcpuFd> for TdxVcpu<'a> {
+    type Error = TdxError;
+
+    fn try_from(fd: &'a mut kvm_ioctls::VcpuFd) -> Result<Self, Self::Error> {
+        // need to enable the X2APIC bit for CPUID[0x1] so that the kernel can call
+        // KVM_SET_MSRS(MSR_IA32_APIC_BASE) without failing
+        let kvm = Kvm::new()?;
```

Comment:
I think we should be re-using the `Kvm` that we initialized the vCPU with. We should include that in this.

`impl<'a> TryFrom<(&'a mut kvm_ioctls::VcpuFd, &'a mut kvm_ioctls::Kvm)> for TdxVcpu<'a>`

Reply:
the `TryFrom` trait only allows one argument from what I can see

Reply:
> the `TryFrom` trait only allows one argument from what I can see

I misread your code snippet, please disregard

Reply:
should be fixed.